comp/core/agenttelemetry/impl/agenttelemetry.go

```diff
@@ -322,38 +327,77 @@ func (a *atel) reportAgentMetrics(session *senderSession, p *Profile) {
 	// Send the metrics if any were filtered
 	a.logComp.Debugf("Reporting Agent Metric telemetry for profile %s", p.Name)
 
-	err = a.sender.sendAgentMetricPayloads(session, metrics)
+	a.sender.sendAgentMetricPayloads(session, metrics)
+}
+
+func (a *atel) loadPayloads(profiles []*Profile) (*senderSession, error) {
+	// Gather all prom metrircs. Currently Gather() does not allow filtering by
+	// matric name, so we need to gather all metrics and filter them on our own.
+	//	pms, err := a.telemetry.Gather(false)
+	pms, err := a.telComp.Gather(false)
 	if err != nil {
-		a.logComp.Errorf("failed to get filtered telemetry metrics: %s", err)
+		a.logComp.Errorf("failed to get filtered telemetry metrics: %v", err)
+		return nil, err
+	}
+
+	session := a.sender.startSession(a.cancelCtx)
+	for _, p := range profiles {
+		a.reportAgentMetrics(session, pms, p)
 	}
+	return session, nil
 }
 
 // run runs the agent telemetry for a given profile. It is triggered by the runner
 // according to the profiles schedule.
 func (a *atel) run(profiles []*Profile) {
-	if a.sender == nil {
-		a.logComp.Errorf("Agent telemetry sender is not initialized")
+	a.logComp.Info("Starting agent telemetry run")
+	session, err := a.loadPayloads(profiles)
+	if err != nil {
+		a.logComp.Errorf("failed to load agent telemetry session: %s", err)
 		return
 	}
 
-	a.logComp.Info("Starting agent telemetry run")
-
-	session := a.sender.startSession(a.cancelCtx)
+	err = a.sender.flushSession(session)
+	if err != nil {
+		a.logComp.Errorf("failed to flush agent telemetry session: %s", err)
+		return
+	}
+}
 
-	for _, p := range profiles {
-		a.reportAgentMetrics(session, p)
+func (a *atel) writePayload(w http.ResponseWriter, _ *http.Request) {
+	if !a.enabled {
+		httputils.SetJSONError(w, errors.New("agent-telemetry is not enabled. See https://docs.datadoghq.com/data_security/agent/?tab=datadogyaml#telemetry-collection for more info"), 400)
+		return
 	}
 
-	err := a.sender.flushSession(session)
+	a.logComp.Info("Showing agent telemetry payload")
+	payload, err := a.GetAsJSON()
 	if err != nil {
-		a.logComp.Errorf("failed to flush agent telemetry session: %s", err)
+		httputils.SetJSONError(w, a.logComp.Error(err.Error()), 500)
 		return
 	}
+
+	w.Write(payload)
 }
 
-// TODO: implement when CLI tool will be implemented
 func (a *atel) GetAsJSON() ([]byte, error) {
```

Comment:
We should probably add a test for this function that validate that the output is scrubbed 😄

Reply:
I will, thank you